Marlin/src/lcd/e3v2/proui/bedlevel_tools.cpp

```diff
@@ -232,46 +233,50 @@ bool BedLevelTools::meshValidate() {
-        isnan(bedlevel.z_values[x][y]) ? COLOR_GREY : (                                                           // gray if undefined
-          (bedlevel.z_values[x][y] < 0 ?
-            (uint16_t)round(0x1F * -bedlevel.z_values[x][y] / rmax) << 11 : // red if mesh point value is negative
-            (uint16_t)round(0x3F *  bedlevel.z_values[x][y] / rmax) << 5) | // green if mesh point value is positive
-              _MIN(0x1F, (((uint8_t)abs(bedlevel.z_values[x][y]) / 10) * 4))),                                    // + blue stepping for every mm
-        start_x_px, start_y_px, end_x_px, end_y_px
+      const float z = bedlevel.z_values[x][y];
+      const uint16_t color = isnan(z) ? COLOR_GREY : (    // Gray if undefined
+        (z < 0 ? uint16_t(round(0x1F * -z / rmax)) << 11  // Red for negative mesh point
+               : uint16_t(round(0x3F *  z / rmax)) << 5)  // Green for positive mesh point
+        | _MIN(0x1F, (((uint8_t)abs(z) / 10) * 4))        // + Blue stepping for every mm
       );
 
+      dwinDrawRectangle(1, color, start_x_px, start_y_px, end_x_px, end_y_px);
+
       safe_delay(10);
       LCD_SERIAL.flushTX();
 
       // Draw value text on
+      if (!viewer_print_value) continue;
+
       const uint8_t fs = DWINUI::fontWidth(meshfont);
-      if (viewer_print_value) {
-        const int8_t offset_y = cell_height_px / 2 - fs;
-        if (isnan(bedlevel.z_values[x][y])) {  // undefined
-          dwinDrawString(false, meshfont, COLOR_WHITE, COLOR_BG_BLUE, start_x_px + cell_width_px / 2 - 5, start_y_px + offset_y, F("X"));
-        }
-        else {                          // has value
-          MString<12> msg;
-          if ((GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10))
-            msg.set(p_float_t(abs(bedlevel.z_values[x][y]), 2));
-          else
-            msg.setf(F("%02i"), uint16_t(abs(bedlevel.z_values[x][y] - int16_t(bedlevel.z_values[x][y])) * 100));
-          const int8_t offset_x = cell_width_px / 2 - (fs / 2) * msg.length() - 2;
-          if ((GRID_MAX_POINTS_X) >= TERN(TJC_DISPLAY, 8, 10))
-            dwinDrawString(false, meshfont, COLOR_WHITE, COLOR_BG_BLUE, start_x_px - 2 + offset_x, start_y_px + offset_y, F("."));
-          dwinDrawString(false, meshfont, COLOR_WHITE, COLOR_BG_BLUE, start_x_px + 1 + offset_x, start_y_px + offset_y, msg);
-        }
-        safe_delay(10);
-        LCD_SERIAL.flushTX();
+      const int8_t offset_y = cell_height_px / 2 - fs;
+      const float z = bedlevel.z_values[x][y];
+      if (isnan(z)) { // undefined
+        dwinDrawString(false, meshfont, COLOR_WHITE, COLOR_BG_BLUE, start_x_px + cell_width_px / 2 - 5, start_y_px + offset_y, F("X"));
       }
-    }
+      else {          // has value
+        MString<12> msg;
+        constexpr bool is_wide = (GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10);
+        if (is_wide)
+          msg.setf(F("%02i"), uint16_t(z * 100) % 100);
+        else
+          msg.set(p_float_t(abs(z), 2));
```

Comment:
@thinkyhead was there a mistake made here?

originally it was like this
```y
    if ((GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10))
      msg.set(p_float_t(abs(z), 2));
    else
      msg.setf(F("%02i"), uint16_t(abs(z - int16_t(z)) * 100));
```
if  
`bool is_wide = (GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10);` 
 
 if that is `TRUE` as the above, wouldnt it remain the same like:  
```y
    if (is_wide)
      msg.set(p_float_t(abs(z), 2));
    else
      msg.setf(F("%02i"), uint16_t(z * 100) % 100);
```


also, when you changed `msg.setf` to get the decimal, I was wondering if that were more efficient. this is the answer I got:

>uint16_t(z * 100) % 100:
>
>This expression multiplies z by 100, converts the result to an unsigned 16-bit integer (uint16_t), and then takes the modulo 100. This approach might be less efficient because it involves a floating-point multiplication and conversion to an integer.

so with that, I'm wondering if avoiding floating-point multiplication is better?

then it says
>However, the actual performance difference might be minimal, and modern compilers are often able to optimize such expressions.